### PR TITLE
performance improvements to flush

### PR DESCRIPTION
**Pull request by yanaise**
Small improvements according to @roironn's request

### CODE REVIEW

**Review by roironn — APPROVED** [review]

**Comment by yanaise** [comment]
Performance improvements to flush

File: packages/core/src/core/plan/plan.ts [hunk 1]
```diff
@@ -160,6 +162,50 @@ const isEqualsNode = async (
...
+  }
+  let currentBefore = await nextDefinedValue(beforeIterator)
+  let currentAfter = await nextDefinedValue(afterIterator)
+  while (!(currentBefore === undefined && currentAfter === undefined)) {
+    if (currentBefore === undefined) {
+      yield { before: undefined, after: currentAfter }
+      // eslint-disable-next-line no-await-in-loop
+      currentAfter = await nextDefinedValue(afterIterator)
+    } else if (currentAfter === undefined
+      || currentBefore.elemID.getFullName() < currentAfter.elemID.getFullName()) {
```
Comment (roironn) [inline packages/core/src/core/plan/plan.ts:191]: I'd probably split into two ifs for readbility

File: packages/core/src/core/plan/plan.ts [hunk 2]
```diff
@@ -221,14 +267,29 @@ const addDifferentElements = (
...
-        topLevelFilters.map(filter => filter(id))
-      )
-    ))
+  const isSpecialId = (id: string): boolean => (BuiltinTypesByFullName[id] !== undefined
+    || elementSource.shouldResolveAsContainerType(id))
+  /**
+   * Ids that represent types or containers need to be handled separately,
+   * because they would not necessary be included in getAll.
+   */
+  const handleSpecialIds = async (elementPair: BeforeAfterElement): Promise<BeforeAfterElement> => {
```
Comment (roironn) [inline packages/core/src/core/plan/plan.ts:276]: I'm 90% sure that all of the "special IDs" should never be included in the diff result - so I think that this entire section could be made much simpler by:
1. Creating a "isContainerTypeID" method in adapter utils
2. Adding an IDFilter which will filter them out.
Thoughts?

File: packages/core/src/local-workspace/remote_map.ts [hunk 3]
```diff
@@ -84,25 +89,33 @@ AsyncIterable<remoteMap.RemoteMapEntry<string>> {
 }
 
 const dbConnections: Record<string, rocksdb> = {}
+const makeId = (length: number): string => {
```
Comment (roironn) [inline packages/core/src/local-workspace/remote_map.ts:92]: Why not use uuid?
Safer and cleaner :)

File: packages/workspace/src/workspace/elements_source.ts [hunk 4]
```diff
@@ -35,41 +35,47 @@ export interface ElementsSource {
   delete(id: ElemID): Promise<void>
 }
 
+export const shouldResolveAsContainerType = (elemName: string): boolean =>
```
Comment (roironn) [inline packages/workspace/src/workspace/elements_source.ts:38]: See comment above

File: packages/workspace/src/workspace/elements_source.ts [hunk 5]
```diff
@@ -35,41 +35,47 @@ export interface ElementsSource {
...
+  (elemName.startsWith(Keywords.LIST_PREFIX)
+    || elemName.startsWith(Keywords.MAP_PREFIX))
+    && (elemName.endsWith(Keywords.GENERICS_SUFFIX))
+
+export const getContainerType = async (fullName: string,
```
Comment (roironn) [inline packages/workspace/src/workspace/elements_source.ts:43]: Why the export? I'm not seeing it used... (Maybe I missed something obvious :))

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts [hunk 6]
```diff
@@ -273,15 +273,15 @@ const buildNaclFilesState = async ({
     deletions: Record<string, Set<T>>,
   ): Promise<void> => {
     const changedKeys = _.uniq(Object.keys(additions).concat(Object.keys(deletions)))
-    return awu(changedKeys).forEach(async key => {
+    return index.setAll(awu(changedKeys).map(async key => {
       const currentValues = (await index.get(key)) ?? []
       const keyDeletionsSet = deletions[key] ?? new Set()
       const keyAdditions = Array.from(additions[key]?.values() ?? [])
       const newValues = currentValues
```
Comment (roironn) [inline packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:280]: This means we have a map with a side effect...
The performance boots definitely worth it - but it should be commented clearly so it won't be missed...

File: packages/workspace/src/workspace/remote_map.ts [hunk 7]
```diff
@@ -95,7 +101,7 @@ export class InMemoryRemoteMap<T, K extends string = string> implements RemoteMa
   }
 
   values(): AsyncIterable<T> {
-    return toAsyncIterable(this.data.values())
+    return toAsyncIterable(Array.from(this.data.keys()).sort().map(key => this.data.get(key) as T))
```
Comment (roironn) [inline packages/workspace/src/workspace/remote_map.ts:104]: I think its a good idea so sort `keys` and `entries` as well

**Review by roironn — APPROVED** [review]
Approved (again :) )
Pending comments. (But plz don't merge until the rebase to lazy intg)

File: packages/core/src/core/plan/plan.ts (hunk 1 above)
Comment (roironn) [reply packages/core/src/core/plan/plan.ts:191]: Also - should probably throw an error if the iterator keys are not sorted ( just throw an error of an iterator key has lower value then the previous key obtained by it )

File: packages/workspace/src/workspace/remote_map.ts (hunk 7 above)
Comment (roironn) [reply packages/workspace/src/workspace/remote_map.ts:104]: Also, using _.sortBy might be better here DYT?

File: packages/core/src/core/plan/plan.ts (hunk 2 above)
Comment (yanaise) [reply packages/core/src/core/plan/plan.ts:276]: That was my initial plan, but it failed the e2e tests. To be honest, I don't really know why this needs to be here. But for now it does. It's not a lot of ids, but yeah maybe some more work is required to figure out what the correct code here is

File: packages/workspace/src/workspace/nacl_files/nacl_files_source.ts (hunk 6 above)
Comment (yanaise) [reply packages/workspace/src/workspace/nacl_files/nacl_files_source.ts:280]: I don't see how the map here has a sideEffect. It's the setAll that has the effect. No? The map just gives us the keys that elements setAll will use

**Pull request closed without merging**